brainiak/reprsimil/brsa.py

```diff
@@ -1659,3 +2460,1350 @@ def _loglike_AR1_singpara(self, param, XTX, XTDX, XTFX, YTY_diag,
+            assert (type(n_nureg) is str and n_nureg == 'BIC') \
+                or (type(n_nureg) is int and n_nureg > 0), \
+                'n_nureg should be a positive number or "BIC"'\
+                'if auto_nuisance is True.'
+        if nureg_method == 'FA':
+            self.nureg_method = lambda x: FactorAnalysis(n_components=x)
+        elif nureg_method == 'PCA':
+            self.nureg_method = lambda x: PCA(n_components=x, whiten=True)
+        elif nureg_method == 'SPCA':
+            self.nureg_method = lambda x: SparsePCA(n_components=x,
+                                                    max_iter=20, tol=tol)
+        elif nureg_method == 'ICA':
+            self.nureg_method = lambda x: FastICA(n_components=x,
+                                                  whiten=True)
+        else:
+            raise ValueError('nureg_method can only be FA, PCA, '
+                             'SPCA(for sparse PCA) or ICA')
+        self.DC_single = DC_single
+        self.logS_range = logS_range
+        self.SNR_bins = SNR_bins
+        self.rho_bins = rho_bins
+        self.tol = tol
+        self.verbose = verbose
+        self.optimizer = optimizer
+        self.rand_seed = rand_seed
+        self.anneal_speed = anneal_speed
+        return
+
+    def fit(self, X, design, nuisance=None, scan_onsets=None):
+        """ Fit the model to data of all participants jointly.
+
+        Parameters
+        ----------
+        X: list of numpy arrays, shape=[time_points, voxels] for each entry.
+            Data to be fitted. Each participant corresponds to one item in
+            the list. If you have multiple scans of the same participants
+            that you want to analyze together, you should concatenate them
+            along the time dimension after proper preprocessing (e.g. spatial
+            alignment), and specify the onsets of each scan in scan_onsets.
+        design: list of numpy arrays, shape=[time_points, conditions] for each.
+            This is the design matrix of each participant.
+            It should only include the hypothetic response for task conditions.
+            You should not include regressors for a DC component or
+            motion parameters, unless with a strong reason.
+            If you want to model head motion, you should include them
+            in nuisance regressors.
+            If you have multiple run, the design matrix
+            of all runs should be concatenated along the time dimension for
+            each participant, with every column for one condition across runs.
+            If the design matrix is the same for all subjects,
+            either provide a list as required, or provide single numpy array.
+        nuisance: optional, list of numpy arrays,
+            shape=[time_points, nuisance_factors] for each subject in the list.
+            Nuisance regressors of each participant.
+            The responses to these regressors will be marginalized out from
```

Comment:
Yes. The variance of the voxel-independent noise is explained by sigma^2. pseudo-SNR and sigma together determines the variance of the task-related signal amplitude beta. But the nuisance regressor variance is not controlled by pseudo-SNR.